File: careerpilot_ui/lib/models/application.dart
class Application {
  final int id;
  final int userId;

  final String jobTitle;
  final String jobCompany;
  final String jobUrl;

  final String? jobLocation;
  final String? jobWorkFormat;
  final String? jobPublishedAt;

  final String jobSource;
  final String jobExternalId;

  final String status;

  final DateTime createdAt;
  final DateTime updatedAt;

  const Application({
    required this.id,
    required this.userId,
    required this.jobTitle,
    required this.jobCompany,
    required this.jobUrl,
    required this.jobLocation,
    required this.jobWorkFormat,
    required this.jobPublishedAt,
    required this.jobSource,
    required this.jobExternalId,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
  });

  String get stableJobKey {
    if (jobSource.isNotEmpty && jobExternalId.isNotEmpty) {
      return '$jobSource::$jobExternalId';
    }

    return jobUrl;
  }

  factory Application.fromJson(Map<String, dynamic> json) {
    return Application(
      id: json['id'] as int,
      userId: json['user_id'] as int,
      jobTitle: json['job_title']?.toString() ?? '',
      jobCompany: json['job_company']?.toString() ?? '',
      jobUrl: json['job_url']?.toString() ?? '',
      jobLocation: json['job_location']?.toString(),
      jobWorkFormat: json['job_work_format']?.toString(),
      jobPublishedAt: json['job_published_at']?.toString(),
      jobSource: json['job_source']?.toString() ?? '',
      jobExternalId: json['job_external_id']?.toString() ?? '',
      status: json['status']?.toString() ?? 'applied',
      createdAt: DateTime.parse(json['created_at'].toString()),
      updatedAt: DateTime.parse(json['updated_at'].toString()),
    );
  }
}

File: apps/api/app/db/models/application.py
from sqlalchemy import (
    Column,
    DateTime,
    ForeignKey,
    Integer,
    String,
    UniqueConstraint,
)
from sqlalchemy.sql import func

from app.db.models.base import Base


APPLICATION_STATUSES = (
    "applied",
    "screening",
    "interview",
    "technical_interview",
    "offer",
    "rejected",
)


class Application(Base):
    __tablename__ = "applications"

    __table_args__ = (
        UniqueConstraint(
            "user_id",
            "job_source",
            "job_external_id",
            name="uq_applications_user_job_identity",
        ),
    )

    id = Column(Integer, primary_key=True, index=True)

    user_id = Column(
        Integer,
        ForeignKey("users.id"),
        nullable=False,
        index=True,
    )

    job_title = Column(String, nullable=False, index=True)
    job_company = Column(String, nullable=False, index=True)
    job_url = Column(String, nullable=False)

    job_location = Column(String, nullable=True)
    job_work_format = Column(String, nullable=True)
    job_published_at = Column(String, nullable=True)
    job_description = Column(String, nullable=True)

    job_source = Column(String, nullable=False, index=True)
    job_external_id = Column(String, nullable=False, index=True)

    status = Column(
        String,
        nullable=False,
        default="applied",
        server_default="applied",
        index=True,
    )

    created_at = Column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )

    updated_at = Column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

    archived_at = Column(
        DateTime(timezone=True),
        nullable=True,
    )
